src/formattedcode/output_cyclonedx.py

```diff
@@ -0,0 +1,659 @@
+            lic.url = declared_license
+        else:
+            lic.name = declared_license
+    # some declared_license entries are also expressed as dicts
+    elif isinstance(declared_license, dict):
+        lic_type = declared_license.get("type")
+        if lic_type in spdx_ids:
+            lic.id = lic_type
+        else:
+            lic.name = lic_type
+        lic.url = declared_license.get("url")
+    return CycloneDxLicenseEntry(license=lic, expression=None)
+
+
+def get_licenses(package: dict) -> List[CycloneDxLicenseEntry]:
+    """map all license data for a package entry to
+    a list of CycloneDX Licenses"""
+    # store previously encountered IDs to identify duplicates
+    seen_ids = set()
+
+    lic_expr = package.get("license_expression")
+    entry = get_license_entry_from_license_expression(lic_expr)
+    if entry is not None:
+        licenses = [entry]
+        if entry.license is not None:
+            seen_ids.add(entry.license.id)
+    else:
+        licenses = []
+
+    declared_license = package["declared_license"]
+    if isinstance(declared_license, list):
+        for entry in declared_license:
+            lic_entry = get_license_entry_from_declared_license(entry)
+            if lic_entry is None:
+                continue
+            id = lic_entry.license.id if lic_entry.license is not None else None
+            if id in seen_ids or id is None:
+                continue
+            else:
+                seen_ids.add(id)
+                if lic_entry is not None:
+                    licenses.append(lic_entry)
+    else:
+        lic_entry = get_license_entry_from_declared_license(declared_license)
+        if lic_entry is not None:
+            id = lic_entry.license.id if lic_entry.license is not None else None
+            if id not in seen_ids and id is not None:
+                seen_ids.add(id)
+                licenses.append(lic_entry)
+
+    return licenses
+
+
+# maps ScanCode URL attributes to CycloneDx external reference types
+url_type_mapping = {
+    "homepage_url": "website",
+    "download_url": "distribution",
+    "bug_tracking_url": "issue-tracker",
+    "code_view_url": "other",
+    "vcs_url": "vcs",
+    "repository_homepage_url": "website",
+    "repository_download_url": "distribution",
+    "api_data_url": "bom"
+}
+
+
+def get_external_refs(package: dict) -> List[CycloneDxExternalRef]:
+    ext_refs = []
+
+    for key in url_type_mapping:
+        ref = get_external_ref_from_key(package, key)
+        if ref is not None:
+            ext_refs.append(ref)
+    return ext_refs
+
+
+def get_hashes_list(package: dict) -> List[CycloneDxHashObject]:
+    hashes = []
+    for alg in hash_type_mapping:
+        if alg in package:
+            digest = package[alg]
+            if alg is not None and digest is not None:
+                hashes.append(CycloneDxHashObject(
+                    hash_type_mapping[alg], digest))
```

Comment:
What about this instead? Also note how I use named keyword arguments which helps if we ever go through some name refactoring and avoid the order dependency implied by positional arguments.

```
    for scancode_algo, cyclonedx_algo in hash_type_mapping.items():
        digest = package.get(scancode_algo)
        if digest:
            hashes.append(CycloneDxHashObject(algo=cyclonedx_algo, content=digest))
```